### PR TITLE
Topic/monadic chain control

### PR DESCRIPTION
**Pull request by EduardSergeev**
Fixes https://github.com/ambiata/disorder.hs/issues/48

### CODE REVIEW

**Comment by markhibberd** [comment]
@EduardSergeev how close is this one?

**Comment by EduardSergeev** [comment]
@markhibberd It works. It isn't perfect (in terms of the implementation), but does what it's supposed to do, so I'll try to prepare it for review this week.

**Comment by EduardSergeev** [comment]
@markhibberd More or less mergable now. Here is a toy example of usage: https://github.com/ambiata/napoleon/pull/123
Definitely not a priority now, but can be useful in future.

**Comment by markhibberd** [comment]
:+1:

File: disorder-fsm/src/Disorder/FSM/Core.hs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import           Control.Monad.Cont (ContT(..))
+import           Control.Monad.IO.Class (MonadIO(..))
+import           Control.Monad.Reader (MonadReader(..))
+import           Control.Monad.State (MonadState(..), evalStateT, modify)
+
+import           Data.Function ((.), ($), const)
+import           Data.Functor (Functor(..))
+import           Data.Maybe (Maybe(..))
+import           Data.Monoid (Monoid(..), (<>))
+import           Data.String (String)
```
Comment (tmcgilchrist) [inline disorder-fsm/src/Disorder/FSM/Core.hs]: Data.Text ?

Comment (EduardSergeev) [inline disorder-fsm/src/Disorder/FSM/Core.hs]: [Eric asked the same question previously](https://github.com/ambiata/disorder.hs/pull/42#discussion_r36814258) And the answer is still the same: these `String`'s are used only in `show` and `fail` so we would still need to convert `Text` to `String` while and with `OverloadedStrings` it shouldn't matter for a user.

Comment (tmcgilchrist) [inline disorder-fsm/src/Disorder/FSM/Core.hs]: Missed that, seems like a reasonable reason.

**Pull request merged**